src/Data/Text/Encoding.hs

```diff
@@ -187,31 +235,311 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
+      w n word8 = len >= n && word8 <= (B.index bs $ len - n)
+      guessUtf8Boundary
+        | w 3 0xf0 = len - 3  -- third to last char starts a four-byte code point
+        | w 2 0xe0 = len - 2  -- pre-last char starts a three-or-four-byte code point
+        | w 1 0xc2 = len - 1  -- last char starts a two-(or more-)byte code point
+        | otherwise = len
 #else
-isValidBS bs = start 0
-  where
-    start ix
-      | ix >= B.length bs = True
-      | otherwise = case utf8DecodeStart (B.unsafeIndex bs ix) of
-        Accept{} -> start (ix + 1)
-        Reject{} -> False
-        Incomplete st _ -> step (ix + 1) st
-    step ix st
-      | ix >= B.length bs = False
-      -- We do not use decoded code point, so passing a dummy value to save an argument.
-      | otherwise = case utf8DecodeContinue (B.unsafeIndex bs ix) st (CodePoint 0) of
-        Accept{} -> start (ix + 1)
-        Reject{} -> False
-        Incomplete st' _ -> step (ix + 1) st'
+  = validateUtf8 0 0 utf8StartState
+    where
 #endif
+      validateUtf8 !ndx0 ndx s
+        | ndx < len =
+          let ndx' = ndx + 1 in
+          case updateUtf8State (B.index bs ndx) s of
+            Just s' ->
+              validateUtf8 (
+                if isUtf8StateIsComplete s'
+                then ndx'
+                else ndx0
+              ) ndx' s'
+            Nothing -> (ndx0, Left $ if ndx == ndx0 then ndx' else ndx)
+        | otherwise = (ndx0, Right $ Utf8ValidState (appendPartialCodePoint 0 $ B.drop ndx0 bs) s)
+
+-- | Validate a 'ByteString' as a contiuation of UTF-8-encoded text.
+--
+-- @validateNextUtf8Chunk chunk s = (n, es)@
+--
+-- This function returns two values:
+--
+-- * The value 'n' indicates the end position of longest prefix of the
+--   'ByteString' that is valid UTF-8-encoded data from the starting
+--   state 's'. If 's' contains an incomplete code point, the input
+--   'ByteString' is considered a continuation. As a result 'n' will be
+--   negative if the code point is still incomplete or is proven to be
+--   invalid.
+--   
+-- * The value 'es' indicates whether the 'ByteString'
+--
+--     * (@Left p@) contains an invalid code point and where the next
+--       (potentially valid) code point begins, so that @p - n@ is the
+--       number of invalid bytes, or
+--     * (@Right s'@) is valid, and all of the remaining bytes starting
+--       at index 'n' are the beginning of an incomplete UTF-8 code
+--       point, and `s'` is the resulting 'Utf8ValidState' value, which
+--       can be used to validate against a following 'ByteString'.
+validateNextUtf8Chunk ::
+#if defined(ASSERTS)
+  HasCallStack =>
 #endif
+  ByteString -> Utf8ValidState -> (Int, Either Int Utf8ValidState)
+validateNextUtf8Chunk bs@(B.length -> len) st@(Utf8ValidState lead s)
+  | len > 0 =
+    let g pos s'
+          -- first things first. let's try to get to the start of the next code point
+          | isUtf8StateIsComplete s' =
```

Comment:
I think the completed codepoint should be part of the output of this function, as a `Maybe Char`.